### PR TITLE
Add support for reading Multi-Release-Jar content

### PR DESCRIPTION
**Pull request by laeubi**
This adds methods to the Jar class to read versioned content of a
multi-release jar for the following cases:

- get a resource or a map of resources for a given release version
- get a merged manifest according to the OSGi specification
- get module name/version for a given given release version
- list all contained release alternative versions

Beside this a unit-test is added to cover these new functions.

Signed-off-by: Christoph Läubrich <laeubi@laeubi-soft.de>

### CODE REVIEW

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java
Comment (timothyjward) [inline biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java:450]: I think that this method isn't quite right or needs a different name and more documentation. At the moment it only checks one version of the release, not every version below, which isn't my understanding of how Multi-Release jars work.

If, for example, my jar has:

    /Foo.class
    /Bar.class
    /META-INF/versions/9/Foo.class
    /META-INF/versions/11/Bar.class

And I call:
*  `getVersionedResource(Foo.class, 11)` - I would expect to get returned the resource at `/META-INF/versions/9/Foo.class`, but this method will return an empty `Optional`
* `getVersionedResource(Bar.class, 9)` - I would expect to get returned the resource at `/Bar.class` but this method will return an empty `Optional`

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java
Comment (timothyjward) [inline biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java:515]: I think that this method has the same issue as `getVersionedResource`, in that it fails to take account of all version folders lower than `release`, with higher versions taking precedence over lower versions.

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java
Comment (timothyjward) [inline biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java:596]: Having checked the specification the location of supplemental manifests is `/META-INF/versions/X/OSGI-INF/MANIFEST.MF`. See https://docs.osgi.org/specification/osgi.core/7.0.0/framework.module.html#framework.module-multireleasejar

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java
Comment (timothyjward) [inline biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java:500]: It looks like we lost a method here, which would break version compatibility.It could just delegate to the new method with version `0`.

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java
Comment (timothyjward) [inline biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java:470]: It looks like we lost a method here which would break version compatibility. It could just delegate to the new method with version `0`.

File: biz.aQute.bndlib/test/aQute/bnd/build/JarTest.java
Comment (timothyjward) [inline biz.aQute.bndlib/test/aQute/bnd/build/JarTest.java:37]: I think this is the wrong path as per the OSGi core spec (link in earlier comments).

**Review by timothyjward — CHANGES_REQUESTED** [review]
Overall I think the direction is good, with just a few small fixes required to match spec behaviour.

I would be interested in seeing the `aQute.bnd.build.model.EE` updated to provide access to the `release` int for each EE.

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java
Comment (laeubi) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java:450]: I can enhance the test-case but actually the method should work that way (but... if its not testes it most likely fails ..).

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java
Comment (laeubi) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java:515]: Do you "think" this or did you tried it out? Just curious, because `headMap` returns a map of all keys lower or equal to release and `lastEntry` returns the highest value from it, sounds exactly as what you explained here...

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java
Comment (laeubi) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java:500]: the method is package private an nowhere used in the package, but if you think it helps, I can add it again...

Comment (timothyjward) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java:500]: I know that removing methods is likely to trigger a baselining error (though default visibility may be safe), and if we see a baselining error it will incur the wrath of @bjhargrave :smile:

Comment (laeubi) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java:500]: Maybe you can enable the workflow runs so we can get a normative result while I address the mentioned issues? :-)

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java
Comment (timothyjward) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java:515]: I don't have a slot to check the code out and run it today but rather than letting the PR sit I did these checks by eyeballing the code. 
It's entirely possible that I've misread the data structure you're using and everything here is actually fine. If that is the case I would recommend some JavaDoc on `getAllVersionMap` so that others don't make the same mistake when maintaining the class in future.

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java
Comment (timothyjward) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java:450]: That would be great - I'm definitely not infallible, and may simply be misunderstanding the role of the internal map structures.

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java
Comment (laeubi) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java:515]: I'll enhance the javadoc and test-case so we can see at least that passing in the CI (it just requires an approval from a bnd committer as I'm a first-time contributor to this repository...).

Comment (laeubi) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java:519]: Done.

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java
Comment (laeubi) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java:470]: Done

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java
Comment (laeubi) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java:500]: Done.

File: biz.aQute.bndlib/test/aQute/bnd/build/JarTest.java
Comment (laeubi) [reply biz.aQute.bndlib/test/aQute/bnd/build/JarTest.java:37]: Fixed.

**Review by laeubi — COMMENTED** [review]
I have applied the suggested fixes and enhanced the test-case.

**Comment by laeubi** [comment]
> I would be interested in seeing the `aQute.bnd.build.model.EE` updated to provide access to the `release` int for each EE.

If I understand the code correctly, this is simply `EE.getCapabilityVersion().getMajor()`

**Comment by timothyjward** [comment]
At this point I'm happy and think this is worth merging - @bjhargrave do you have anything to add?

**Comment by timothyjward** [comment]
> If I understand the code correctly, this is simply `EE.getCapabilityVersion().getMajor()`

Yes, although the following isn't exactly readable...

    EE myEE = getResolveTargetEE();
    manifest = jar.getManifest(myEE.getCapabilityVersion().getMajor()).get();

**Review by timothyjward — APPROVED** [review]

**Comment by laeubi** [comment]
I think it won't be a problem to add a new "getRelease()" method here, but lets see first if there is other to do.

**Comment by laeubi** [comment]
thanks @timothyjward for the review, the Ci is also happy!

**Comment by laeubi** [comment]
> Yes, although the following isn't exactly readable...
> 
> ```
> EE myEE = getResolveTargetEE();
> manifest = jar.getManifest(myEE.getCapabilityVersion().getMajor()).get();
> ```

I have opened 
- https://github.com/bndtools/bnd/pull/5358

for this as it seems to be a valid enhancement regardless of MR support...

**Comment by laeubi** [comment]
@bjhargrave I have now created two PRs based on this to show how one can uses the Multi-Release-Jar support to:

- Correctly handle JPMS module info: https://github.com/bndtools/bnd/pull/5360
- Create a Multi-Release-Bundle by specify `Multi-Release: true` in the BND instructions: https://github.com/bndtools/bnd/pull/5359

I hope this i sufficient to review this PR and decide if changes are required before this can be merged.

FYI @stbischof @rotty3000

**Comment by timothyjward** [comment]
The following comment is wrong. The version should be 17 for the third check in the test. 

https://github.com/bndtools/bnd/compare/2d1a4c6ba66197105067b44814b5c4033f6f7a0d..9856c6cbda183da1a08ee1a8fd1dbaa77a76e090#diff-5ba9b2d89cbd966cfd1655fb0c9488599e851bc6ae563e20bfd63c9df4e0d93aR152

**Comment by laeubi** [comment]
> The version should be 17 for the third check in the test.

Fixed!

**Comment by timothyjward** [comment]
@bjhargrave - I think that this should be merged, can we go ahead?

**Comment by laeubi** [comment]
@rotty3000 can you maybe help here? Its about a full month now with no feedback at all ... :-\

**Comment by bjhargrave** [comment]
I am still working on ideas about mrjar support in a side branch based upon the narrative in https://github.com/bndtools/bnd/wiki/Multi-release-JAR-support-design-discussion. Sorry it is taking awhile but have other items to also work on.

Until I have a better idea on how we should handle mrjars, I don't want to merge this partial proposal.

**Comment by laeubi** [comment]
So now after about half a year later is there any decision / progress on this?

**Comment by pkriens** [comment]
@laeubi can you contact me via Peter.Kriens@aQute.biz? I'd like to have a zoom call to understand all this. I've kept myself out of this discussion (I am not convinced multi release jars are even remotely a good idea, they seem to suffer a lot of accidental complexity) but now with BJ out of the loop I need to get this of the PR list. So if you can contact me, we setup a call.

**Comment by laeubi** [comment]
@pkriens I send you a mail, I think the main point is not if one likes MR or not, that's not a desiccation made by the tool, I just wanted to support it because more and more libs are start using this and it is part of the OSGi Spec for many years, but yes it might not be used much. That's also why I aim for a quite basic solution as a first step and from my testing with this it works quite good for the rare cases where I needed a MR jar (mostly to add OSGi headers to a 3rd party library).

**Comment by pkriens** [comment]
I will take over this PR

**Pull request closed without merging**
